**A simple A <-> B reaction between 2 species with initial uniform concentrations,
with 1st-order kinetics in both directions, taken to equilibrium**

Diffusion (non-applicable) not done

In [1]:
from modules.chemicals.chemicals import Chemicals as chem
from modules.reactions.reactions import Reactions
from life_1D.bio_sim_1d import BioSim1D as bio

from modules.html_log.html_log import HtmlLog as log



# Initialize the system
chem_data = chem(diffusion_rates=[0.1, 0.1], names=["A", "B"])  # diffusion_rates NOT used for now
bio.initialize_system(n_bins=3, chem_data=chem_data)

bio.set_uniform_concentration(species_index=0, conc=10.)
bio.set_uniform_concentration(species_index=1, conc=50.)

bio.describe_state(show_diffusion_rates=True)

3 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [10. 10. 10.]
  Species 1 (B). Diff rate: 0.1. Conc:  [50. 50. 50.]


In [2]:
rxn = Reactions(chem_data)

# Reaction A -> B , with 1st-order kinetics in both directions
rxn.add_reaction(reactants=["A"], products=["B"], forward_rate=3., reverse_rate=2.)

bio.all_reactions = rxn

print("Number of reactions: ", rxn.number_of_reactions())

Number of reactions:  1


In [3]:
rxn._internal_reactions_data()    # Low-level view of the reactions data

0: [(1, 0, 1)] <-> [(1, 1, 1)]   ; Fwd: 3.0 / Back: 2.0


In [4]:
rxn.describe_reactions()

Number of reactions:  1
0: A <-> B  (Rf = 3.0 / Rb = 2.0)


['0: A <-> B  (Rf = 3.0 / Rb = 2.0)']

In [5]:
# First step
bio.react(time_step=0.1, n_steps=1)
bio.describe_state()

3 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [17. 17. 17.]
  Species 1 (B). Diff rate: 0.1. Conc:  [43. 43. 43.]


NOTE: the concentration of species A is increasing, while that of species B is decreasing.
All bins have identical concentrations; so, there's no diffusion:
[[17. 17. 17.]
 [43. 43. 43.]]

In [6]:
# Numerous more steps
bio.react(time_step=0.1, n_steps=10)

bio.describe_state()

3 bins and 2 species:

  Species 0 (A). Diff rate: 0.1. Conc:  [23.99316406 23.99316406 23.99316406]
  Species 1 (B). Diff rate: 0.1. Conc:  [36.00683594 36.00683594 36.00683594]


NOTE: Consistent with the 3/2 ratio of forward/reverse rates (and the 1st order reactions),
 the systems settles in the following equilibrium:

[A] = 23.99316406
 
[B] = 36.00683594


In [7]:
A_eq = bio.bin_concentration(0, 0)
B_eq = bio.bin_concentration(0, 1)
print(f"Ratio of equilibrium concentrations: {B_eq / A_eq}")
print(f"Ratio of forward/reverse rates: {rxn.get_forward_rate(0) / rxn.get_reverse_rate(0)}")

Ratio of equilibrium concentrations: 1.5007122797020636
Ratio of forward/reverse rates: 1.5
